cpp/src/arrow/io/file.cc

```diff
@@ -152,20 +152,30 @@ static inline int64_t lseek64_compat(int fd, int64_t pos, int whence) {
 #endif
 }
 
+#if defined(_MSC_VER)
+static inline Status ConvertToUtf16(const std::string& input, std::wstring* result) {
+  if (result == nullptr) { return Status::Invalid("Pointer to result is not valid"); }
+
+  if (input.empty()) {
+    *result = std::wstring();
+    return Status::OK();
+  }
+
+  std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> utf16_converter;
+  *result = utf16_converter.from_bytes(input);
```

Comment:
It seems this can fail either with an `std::range_error` or a decoding error set in the `utf16_converter.state()`:

https://msdn.microsoft.com/en-us/library/ee191684.aspx#Anchor_5

Might be worth writing a unit test to trigger these cases